File: src/Web/JavaScriptJS/patterns/object_pool.js
const assert = require("node:assert/strict");
/** Reuses bounded expensive objects instead of creating one per request. */
function run() {
  const available = [{ id: 1 }, { id: 2 }];
  const pool = { acquire: () => available.pop(), release: (item) => available.push(item) };
  const item = pool.acquire();
  assert.equal(available.length, 1);
  pool.release(item);
  assert.equal(available.length, 2);
  return "object-pool";
}
module.exports = { run };
if (require.main === module) run();

File: src/Scripting/Lua/patterns/object_pool.lua
local pool = { free = { {id=1}, {id=2} }, used = {} }
function pool:acquire()
  local item = table.remove(self.free)
  self.used[item.id] = item
  return item
end
function pool:release(item)
  self.used[item.id] = nil
  table.insert(self.free, item)
end
local item = pool:acquire()
assert(pool.used[item.id] == item)
pool:release(item)
assert(pool.used[item.id] == nil and #pool.free == 2)
return true
